oap-server/server-query-plugin/query-graphql-plugin/src/main/java/org/apache/skywalking/oap/query/graphql/resolver/MetricQuery.java

```diff
@@ -59,17 +64,28 @@ public IntValues getValues(final BatchMetricConditions metrics, final Duration d
             kv.setValue(query.readMetricsValue(condition, duration));
             values.addKVInt(kv);
         } else {
-            for (final String id : metrics.getIds()) {
-                KVInt kv = new KVInt();
-                kv.setId(id);
+            List<CompletableFuture> futureList = new LinkedList<>();
+            metrics.getIds().stream().forEach(id -> {
 
                 MetricsCondition condition = new MetricsCondition();
                 condition.setName(metrics.getName());
                 condition.setEntity(new MockEntity(id));
-
-                kv.setValue(query.readMetricsValue(condition, duration));
-                values.addKVInt(kv);
-            }
+                CompletableFuture future = CompletableFuture.supplyAsync(() -> {
```

Comment:
The question is, what exactly the case you are facing to make you feel you need query to be async? One way or another, UI has to wait for a database response.

Reply:
This means that the id is queried concurrently, and you don't have to wait for each result before proceeding the next time.

Reply:
I think the best way is to modify MetricsQueryEsDAO to support batch query of ids

Reply:
Yes, but v1 metric APIs have been `Deprecated`, we recommend using v2, and making this kind of bulk questions as different GraphQL level queries.

Reply:
When there are many applications, the topology map display is too slow. The reason is that in the previous position, each id triggers a query, and the next query needs to wait for the query result of the previous id.

Reply:
Which version's UI are you talking about?

Reply:
v8.9.1